### ISSUE TITLE
Upload pump status and treatment info to NS

### ISSUE THREAD

**Issue by Pogman** [pazaan/600SeriesAndroidUploader#147]
Update NS with pump status and treatments.

Graph:

Bolus : normal square dual
Basal: temp rate/percentage
Suspend / Resume (treated as a temp basal at 0u for NS accounting)
CAGE: time of cannula/reservoir change
SAGE: time of sensor change
Pump battery: drop note to NS when change detected

Status:

Active alert on pump
CGM status (offline, warmup, calibrating)
CGM transmitter battery
CGM calibration status
CGM time to next calibration
Dual/Square delivery status (units, time remaining)
Temp basal delivery status (units, time remaining)

**Comment by Pogman** [comment]
Preliminary testing good...

![ns-bolus](https://user-images.githubusercontent.com/7413594/28794746-46d159fa-762f-11e7-9db7-7aab8fcf3bf9.JPG)
![ns-bolus-reports](https://user-images.githubusercontent.com/7413594/28794751-4b8a5258-762f-11e7-967a-7a1e4e49925c.JPG)

**Comment by pazaan** [comment]
Is this in a branch somewhere? `pogman/bolus-info`?

**Comment by Pogman** [comment]
Not yet - it's a bit of a construction zone right now.

**Comment by Pogman** [comment]
Status of current bolus mode and delivery. Will update after each poll.

![ns-bolus-status](https://user-images.githubusercontent.com/7413594/28821435-489dceb6-76ad-11e7-9fe3-8169b3105241.JPG)

**Comment by Pogman** [comment]
Status + time until next calibration

![ns-status-cal](https://user-images.githubusercontent.com/7413594/28882146-498d9a2c-77a2-11e7-8d21-26bccaf03a90.JPG)

**Comment by Bal00** [comment]
Nice work! Do you have access to the carbohydrate entered into bolus wizard?

**Comment by Pogman** [comment]
Unsure as I don't use the wizard personally but there is a few unknowns left in the message that's read from the pump and those unknowns are maybe for the wizard or the smartgaurd status. 

I've fairly heavily mined the status message at this stage though...

```
----------------------------------------------------------

Status Message Payload

           ?0 ?1 ?2 ?3 ?4 ?5 ?6 ?7 ?8 ?9 ?A ?B ?C ?D ?E ?F
0x00000000 XX XX XX QQ nb NB NB NB -- -- -- -- NM NM NR nr
0x00000010 kb KB KB KB KT KT KT KT KR kr BP BR BR BR BR tb
0x00000020 TB TB TB TR TM TM bu BU BU BU PP rr RR RR RR RH
0x00000030 RM ii II II II SV SV ST ST ST ST SO SO SO SO LS
0x00000040 SD SS SU SC SC SB SX SX BW BG BG WW WW WT WT WT
0x00000050 WT WO WO WO WO -- -- -- sm sm SM SM sn sn SN SN

XX = header

QQ = pump status
BP = basal pattern
BR = basal rate (div 10000)
TB = temp basal (div 10000)
TR = temp basal percent
TM = temp minutes remaining
BU = basal units delivered today (div 10000)
PP = battery
RR = reservoir (div 10000)
RH = reservoir hours remaining
RM = reservoir mins remaining
II = active insulin (div 10000)
SV = sgv
ST = sgv time rtc
SO = sgv time offset
LS = low suspend active
SD = sgv trend
BW = bolus wizard
BG = bgv
WW = alert
WT = alert time rtc
WO = alert time offset

NB = now bolusing amount delivered (div 10000)
NM = now bolusing minutes remaining
NR = now bolusing reference

KB = last bolus amount (div 10000)
KT = last bolus time
KR = last bolus reference

SS = sensor status
SU = sensor control?
SC = sensor calibration minutes remaining
SB = sensor battery
SX = sensor rate of change (div 100)

SM = sensor mode active (incs each on and each off, if bit 0 is set then on)
SN = sensor mode active (incs each on and each off, if bit 0 is set then on)

-- = unknown use or zero while monitoring

lower case used for unconfirmed but assumed related values ie: rr RR RR RR (is it 24bit or 32bit) 

----------------------------------------------------------
```

**Comment by Pogman** [comment]
It will be in the history message but dealing with the history in real time and not disturbing cgm radio etc will take some consideration.

**Comment by cfaagaard** [comment]
That looks so good. How did you get the extra info into the pump-field? I have tried to get extra info into the "screen" as well, but no much luck.

**Comment by cfaagaard** [comment]
About history in real time: I have been testing different settings in my project when reading the history. And for example if you look 2 days back it will add an extra 10-15 seconds to your 5-minutes-interval-loop to handle the data.

**Comment by Pogman** [comment]
The mLab data should show you what you need:

```
{
    "_id": {
        "$oid": "598e33af05abc83c17a3a353"
    },
    "created_at": "2017-08-11T23:45:49+0100",
    "device": "medtronic-600://6213-1102475",
    "pump": {
        "battery": {
            "percent": 75
        },
        "clock": "2017-08-11T23:45:49+0100",
        "iob": {
            "bolusiob": 4.2,
            "timestamp": "Aug 11, 2017 11:45:49 PM"
        },
        "reservoir": 44,
        "status": {
            "bolusing": false,
            "status": "⚠ dual>>6.975u-51m :::: calibrate now!",
            "suspended": false
        }
    },
    "uploaderBattery": 35
}
```

Just write to the status var and set suspended and bolusing to false for it to show in NS.

BTW this particular status shows an active alert, dual bolus amount delivered with time remaining, sensor battery (full = :::: and this bar reduces with use), calibration status or time to next calibration.

There will be a code branch up here in a few days with all this pump info pushed to NS.

**Comment by Pogman** [comment]
Did you do your history pull with EHSM or without?

We can't use EHSM for speedup as it causes the pump to change radio channel and depending on the radio environment this can be infrequent to several times an hour causing cgm loss for 10-30 minutes.

**Comment by Pogman** [comment]
Status info

![ns-status-dual](https://user-images.githubusercontent.com/7413594/29235429-a02aa478-7ef6-11e7-88e4-265376374911.JPG)

![ns-status-0u 100 alert normal sb 2h40m](https://user-images.githubusercontent.com/7413594/29235438-b71cdca0-7ef6-11e7-9e41-63be75e4dc76.JPG)

**Comment by Pogman** [comment]
Full NS screen showing state after a sensor change, through warmup (no SGV but there are continuous updates to NS of BG + Bolus etc), after a sensor calibration I see SG rise and bolus then my running dual finishes and sends a notification to NS and updates the basal (NS shows a combo bolus by bumping the basal graph) - finally all is normal with no alerts, full sensor battery and 5 hours or so to next calibration.

The SAGE time pill is set to time of sensor change automatically. Also CAGE will be set automatically after a reservoir change and when pump battery is changed a note is pushed to NS too. 

![ns-state-1](https://user-images.githubusercontent.com/7413594/29235669-a78282c0-7ef8-11e7-8d9f-a80907321328.JPG)

**Comment by cfaagaard** [comment]
I am using the same "communication" settings as you for getting the pumptime/pumpstatus. I have just put in the history request/response also. I have not observed any loss of CGM and for the last 48 hours it has used the same radio channel in the 5-minute-interval-loop. 
But I will be aware of this, thanks.

**Comment by cfaagaard** [comment]
Thanks for status-json.

**Comment by HenrikHL** [comment]
@Pogman is this an unreleased version of the uploader showing lots of icons on the nightscout site (I just have the green/yellow/red BS -dots? When is this expected to be released?
Latest (0.5.1) release does show boluses from the Pump - but nothing else. Would really like to know if the Pump has canceled a Combined Bolus and see if the Basal is above or below 100%...

**Comment by Pogman** [comment]
@HenrikHL it will probably be some time yet to an official release as it's still very much WIP. If you can build yourself you can pull the code and play with that. There will be a code branch up here soon or if your feeling brave you can pull the code from my repo.

**Comment by HenrikHL** [comment]
Thnak you @Pogman - I will build it myself but will wait for you to make a branch here ;-)

Really looking forward to this - I have been looking for this a long time and thought my Nightscout setup was wrong...

**Comment by HenrikHL** [comment]
Btw - will it require any setup of CarePortal in Nightscout - or is all information available in the JSON data sent

**Comment by Pogman** [comment]
If you want to see all the available pump data you would need these settings in Azure:

```
Connection strings

ENABLE
careportal cage sage basal pump iob bwp

DEVICESTATUS_ADVANCED
true

PUMP_FIELDS
reservoir battery status

SHOW_PLUGINS
careportal cage sage basal pump iob bwp
```

**Comment by HenrikHL** [comment]
I have all of the above already (though I am using IAGE instead of CAGE)
I have not enabled BWP - as I could not figure out how to use this + enabling it didn't seem to provide any new info in NightScout
DEVICESTATUS_ADVANCED=true - I was not aware of this setting. I will enable it in Azure tonight...

**Comment by Pogman** [comment]
You probably don't need BWP and it's not something that is written to here. There was something in the NS docs about having this set if Careportal is being used but that could be outdated info. NS docs also say that DEVICESTATUS_ADVANCED needs to be set for pump status stuff to show.

**Comment by Bal00** [comment]
Please may I suggest / request a switch in settings for sending treatment data? If a user is looping with a different pump at night, two sets of temp basals would be uploaded which would be problematic. Other info (SGV, BG) from the 600 pump are still relevant while looping.

**Comment by Pogman** [comment]
Suggestions most welcome - keep em coming.

**Comment by Pogman** [comment]
Notes for documenting treatments feature:
```
-----------------------------------------------------------------------------------------------

600 SERIES UPLOADER:
open settings
toggle "Enable upload of treatments to Nightscout"

-----------------------------------------------------------------------------------------------

TREATMENTS:
nightscout chart and reports / CAGE SAGE BASAL pills 

BOLUS
precise delivery start time 

DUAL BOLUS
precise delivery start time

SQUARE BOLUS
precise delivery start time

TEMP BASAL
approx started time (range)
approx ended time (range)

SUSPEND
approx started time (range)
approx ended time (range)

SAGE (sensor age)
approx changed time (range)

CAGE (cannula/reservoir  age) 
approx changed time (range)

BATTERY NOTE (pump battery change)
approx changed time (range)

-----------------------------------------------------------------------------------------------

STATUS:
precise pump/cgm status at last poll (nightscout Pump pill)

[reservoir] [battery] [age] [alert] [mode] [cgm] [calibration]

[reservoir]
units of insulin remaining

[battery]
pump battery percent remaining

[age]
age of status data in minutes ago.

[alert]
active alert on pump

[mode]
pump insulin delivery mode

normal
suspended
bolusing
dual>>[units delivered]-[time remaining]
square>>[units delivered]-[time remaining]
temp>>[units]-[time remaining]
temp>>[percent]-[time remaining]

[cgm]
cgm n/a (pump is not receiving data from cgm/sensor)
cgm transmitter battery state (  ::::  :::.  ::..  :...  ....  )

[calibration]
time to next calibration (hours and minutes)
warmup
calibrating
cal.complete
calibrate now!

-----------------------------------------------------------------------------------------------

NIGHTSCOUT
azure "Connection strings" needed

ENABLE
careportal cage sage basal pump iob

DEVICESTATUS_ADVANCED
true

PUMP_FIELDS
reservoir battery clock status

SHOW_PLUGINS
careportal cage sage basal pump iob

-----------------------------------------------------------------------------------------------
```

**Comment by Pogman** [comment]
Branch added for testing:

https://github.com/pazaan/600SeriesAndroidUploader/commits/develop-treatments

This should be quite solid for production use. Go at it - feedback, gotchas, you killed my baby you monster welcome!

**Comment by SJoWie80** [comment]
Do you have an .apk so i can test this ?

**Comment by Pogman** [comment]
@SJoWie80 you will need to be able to build this yourself to test.

**Comment by SJoWie80** [comment]
@Pogman Finaly got around to build and use it.  Now active for a week works great!

**Comment by whooze** [comment]
Sounds good, I'm SO gonna try to build this myself and give it a try too

2017-09-29 11:01 GMT+02:00 SJoWie80 <notifications@github.com>:

**Comment by Tausta** [comment]
Great work! It works as it should be.

Nightscout web page version 0.9.2 (latest release) does not work if there is no sensor value in the last 48 hours. Perhaps the current development version would work better? Also temp basal rates are displayed correctly in the Nightscout, but normal basal is not.
 
We typically do not have the uploader all the time close to the pump. That’s why there is gaps. It would be great if the uploader could retrieve missing data from the history.

**Comment by Pogman** [comment]
Thanks guys for testing and feedback.

Nightscout does not really like to work without CGM data and running with just finger bgs without sensor readings causes the graph to act up trying to stay synced to a last CGM date. A potential workaround for this is an advanced option in the uploader to insert BG's as CGM points in NS.

Pump basal patterns and pump history handling are planned though from a project pov structurally different. This is just the real-time portion and is designed to have no addition battery stress on uploader device or pump.

**Comment by Pogman** [comment]
Guys - "treatments" branch is depreciated, please try out the all new and shiny "Indra" - This does all your history in realtime so precise times for everything and bolus wizard data such as carbs etc. Also backfills anything thats missing so no cgm gaps. Also pump profiles for basal patterns and pattern changes. If you use the uploader for yourself there is support for the Urchin pebble watchface, this works offline so no need for a network and fast updates of data, also lots of extended pump data on the watchface fully customisable. Latest release version of NS is recommended for pump profile use.

**Comment by diamolly** [comment]
I just ran into something Id like to share. First things first: WOW Pogman! Indra is amazing. I appreciate all the hard work you put into this. You are making my 4 year olds life a much better one  because of these features :)

To the issue…
I have two uploader phones, both running Indra now. When I install the Indra-version and plug the CNL in it works fine. When the phones battery is done I plug the CNL into my secondary phone - works great!

But once I recharge my first phone and reconnect the CNL I get this error:

“Error uploading: length=189; Index=189”

The phone is receiving all the info (SGV, history graph) just not uploading it to nightscout. In order to solve this, I have to wipe all info from the app or reinstall it. 

Any ideas?

**Comment by Pogman** [comment]
@diamolly hi thanks for testing - this has something to do with profiles and potentially a difference in what is being sent from the pump from what is expected currently. Once you update a profile it's going to get caught in this error and not upload anything else. For now you can avoid this by not initiating a profile update. We're looking into it and might need to get your raw profile data to understand what is different.

**Comment by Pogman** [comment]
@diamolly actually found the cause of this issue - there will be an update tomorrow

**Comment by Pogman** [comment]
Branch updated with fixes and tweaks. If you were having a error uploading you will need to clear the app data after building this version to clear out the bad data.

@diamolly please let me know how you get on with this - really want to hear from users that run multiple uploaders

**Comment by diamolly** [comment]
Thanks! I´ll let you know how this version works, building now :)

**Comment by diamolly** [comment]
@Pogman After the last branch update I have not had the same issue again. But I am getting a new error now:

«Error uploading: Unable to resolve host ‘MYSITE.azurewebsites.net’: No address associated with hostname»

Im only getting this error on one of my uploaders. Here are my uploaders:

Moto G, android 6.0.1 - No errors
Xeno J5, android 5.1 - Upload error

The error does not occur every poll, several successfull uploads can happen before it suddenly gets the error. And after a few new polls it sometimes gets back to uploading again.

Im going to try another uploader i have (sony) to see if im getting the issue there aswell.

**Comment by j144** [comment]
Great work! No issues with Moto G (thea) LineageOS 7.1.2. 
I think we can now use autotune: http://openaps.readthedocs.io/en/latest/docs/Customize-Iterate/autotune.html#phase-c-running-autotune-for-suggested-adjustments-without-an-openaps-rig

**Comment by Pogman** [comment]
@diamolly that error means there is no network and can't find that NS address. You will get that error if you lose wifi or mobile connection.

**Comment by Pogman** [comment]
@j144 welcome to the party - there's a lot going on in Indra so more folk kicking it around the better!

**Comment by HenrikHL** [comment]
I have tried to build this a couple of times but every time I try to install the .apk on my Galaxy (A3?) it exists - there is no problem using the "official" .apk-version

Probably something I am doing wrong when I build the .apk...?!? Btw - does it need to be signed?

**Comment by Pogman** [comment]
@HenrikHL you will need to completely uninstall the official release when you put your own built version on the phone

**Comment by HenrikHL** [comment]
ahhh - thanx

**Comment by Pogman** [comment]
You probably do need to have it signed with your own generated key too for it to install. Been so long since I've touched any of that stuff but I do remember it was one of the first things I needed to do when starting android dev.

**Comment by HenrikHL** [comment]
Thanx @Pogman this is fantastic. It has intalled the APP - it works fantastic!!! Great work! All treatments and rates uploaded to Nightscout. We normally only run the APP during the night - but this gives us the full overview

**Comment by Pogman** [comment]
Cool - good to hear. You can dig into the settings and get your past data upto 90 days worth too. Also it's worth adding some settings to show the extra pump status info:
![indra-connection-strings](https://user-images.githubusercontent.com/7413594/34469910-2936eacc-ef1f-11e7-82ee-05da59020584.JPG)

**Comment by Alexhaf** [comment]
Happy New Year and Thumbs up, @Pogman  !!!
We love Indra :-) Testing it on two uploaders (Samsung Galaxy S5) and no issues so far today. 
I will install a third one later tonight (we need an uploader on each floor for our 4 years old T1D)  :-)
Keep up the good work!!

**Comment by Pogman** [comment]
@Alexhaf woah hardcore... 3 uploaders. Are you running them simultaneously?

**Comment by Alexhaf** [comment]
yes, simultaneously :-)

**Comment by Pogman** [comment]
Ok - now this should work ok but may be better to slightly offset the polling on each one for best results and to stop clashing. You can do this via the "advanced system settings" and  "poll timing" change the grace period so A is 30 sec B is 60 sec and C is 90 sec - though you can experiment but these would be good values I think.

**Comment by Alexhaf** [comment]
Thank you, I will change the settings..

**Comment by Pogman** [comment]
closing - feature available on 0.6+
